sycl/include/sycl/vector_preview.hpp

```diff
@@ -779,36 +462,39 @@ template <typename Type, int NumElements> class vec {
   template <typename convertT,
             rounding_mode roundingMode = rounding_mode::automatic>
   std::enable_if_t<
-      !std::is_same_v<vec_data_t<DataT>, vec_data_t<convertT>> &&
-          !std::is_same_v<detail::ConvertToOpenCLType_t<vec_data_t<DataT>>,
-                          detail::ConvertToOpenCLType_t<vec_data_t<convertT>>>,
+      !std::is_same_v<element_type_for_vector_t<DataT>, element_type_for_vector_t<convertT>> &&
+          !std::is_same_v<detail::ConvertToOpenCLType_t<element_type_for_vector_t<DataT>>,
+                          detail::ConvertToOpenCLType_t<element_type_for_vector_t<convertT>>>,
       vec<convertT, NumElements>>
   convert() const {
-    static_assert(std::is_integral_v<vec_data_t<convertT>> ||
+    static_assert(std::is_integral_v<element_type_for_vector_t<convertT>> ||
                       detail::is_floating_point<convertT>::value,
                   "Unsupported convertT");
-    using T = vec_data_t<DataT>;
-    using R = vec_data_t<convertT>;
+    using T = typename std::conditional_t<std::is_same_v<DataT, sycl::half>, DataT, element_type_for_vector_t<DataT>>;
+    using R = typename std::conditional_t<std::is_same_v<convertT, sycl::half>, convertT, element_type_for_vector_t<convertT>>;
     using OpenCLT = detail::ConvertToOpenCLType_t<T>;
     using OpenCLR = detail::ConvertToOpenCLType_t<R>;
     vec<convertT, NumElements> Result;
 
-#if defined(__SYCL_DEVICE_ONLY__)
+#ifdef __SYCL_DEVICE_ONLY__
     using OpenCLVecT = OpenCLT __attribute__((ext_vector_type(NumElements)));
     using OpenCLVecR = OpenCLR __attribute__((ext_vector_type(NumElements)));
+
+    auto NativeVector = VecConvertToOpenCLType<DataType, vector_t>(m_Data);
```

Comment:
```suggestion
    auto NativeVector = VecConvertToOpenCLType(m_Data);
```

Should just work here and in other places if we decide to keep this helper (see my other comment on that).